torch/csrc/jit/passes/common_expression_hoisting.cpp

```diff
@@ -0,0 +1,111 @@
+#include <torch/csrc/jit/passes/common_expression_hoisting.h>
+
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/ir/node_hashing.h>
+#include <torch/csrc/jit/jit_log.h>
+
+#include <cstddef>
+#include <unordered_set>
+
+namespace torch {
+namespace jit {
+namespace {
+
+bool HoistFromIfNode(Node* if_node, AliasDb& aliasDb) {
+  Block* true_block = if_node->blocks()[0];
+  Block* false_block = if_node->blocks()[1];
+  // find common statements in the two subblocks
+  bool block_modified = false;
+
+  auto true_block_nodes = std::unordered_set<Node*, HashNode, EqualNode>(
+      true_block->nodes().begin(), true_block->nodes().end());
+  for (auto it = false_block->nodes().begin();
+       it != false_block->nodes().end();) {
+    Node* false_b_node = *it;
+    // node may be moved to a different block so advance iterator now
+    ++it;
+
+    auto matching_elem = true_block_nodes.find(false_b_node);
+    if (matching_elem == true_block_nodes.end()) {
+      continue;
+    }
+    Node* true_b_node = *matching_elem;
+
+    // Check if there are any mutatiors for the output blocks
+    // Technically you only need to check for it in the IF block
+    if (aliasDb.hasOutputWriters(false_b_node) ||
+        aliasDb.hasOutputWriters(true_b_node)) {
+      continue;
+    }
+
+    // Check if a move to the front of the block is valid
+    // If both of the moves are valid, then we know we can move the item out of
+    // the if blocks entirely.
+    bool true_moveable = aliasDb.couldMoveBeforeTopologically(
+        true_b_node, true_block->nodes().front());
+    bool false_moveable = aliasDb.couldMoveBeforeTopologically(
+        false_b_node, false_block->nodes().front());
+
+    if (!true_moveable || !false_moveable) {
+      continue;
+    }
+    // Now do the hoist”
+    block_modified = true;
+    false_b_node->moveBefore(if_node);
+    true_b_node->replaceAllUsesWith(false_b_node);
+
+    // fix up the if block outputs
+
+    true_block_nodes.erase(true_b_node);
```

Comment:
I don't understand why that is necessary.

Reply:
we initialize `true_block_nodes` at the start of the function, so i think anytime we move a node out of the if node and into it's surrounding block we are invalidating the hash of all nodes in `true_block_nodes` which have as input any outputs of the newly moved node

Reply:
this sounds important. It sounds like we will fail to hoist out any expressions whose inputs have already been hoisted out.

Reply:
Yeah, didn't think about how the hash would change. 

Is it better to remove the items, and then readd them when we change the inputs, or to just remove the set completely and just loop over the if block (of course the latter has a higher big-O complexity)

Reply:
(also was able to repro this failure with an additional test :) )

Reply:
Yes, the complexity with removing and re adding is O(#num changes) * O(# avg number of node users in block) (typically 1-2) instead of O(#num changes) * O(# of nodes in block)